scripts/src/java/org/oppia/android/scripts/XMLSyntaxCheck.kt

```diff
@@ -0,0 +1,71 @@
+package org.oppia.android.scripts
+
+import org.xml.sax.SAXParseException
+import java.io.File
+import javax.xml.parsers.DocumentBuilder
+import javax.xml.parsers.DocumentBuilderFactory
+
+/**
+ * Script for ensuring that all the XML files
+ * in the repo are syntactically correct.
+ */
+class XMLSyntaxCheck {
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
+      // path of the repo to be analyzed.
+      val repoPath = args[0] + "/"
+
+      // a list of all allowed directories in the repo to be analyzed.
+      // args[0] is the repoPath, the allowed directories are specified
+      // after it. Hence, we have to start from the 1st index.
+      val allowedDirectories = args.drop(1)
+
+      // a list of all XML files in the repo to be analyzed.
+      val searchFiles = RepoFile.collectSearchFiles(
+        repoPath,
+        allowedDirectories,
+        ".xml"
+      )
+
+      // class object which is needed to acess the helper methods
+      val xmlSyntaxCheck: XMLSyntaxCheck = XMLSyntaxCheck()
+
+      // builder factory which provides the builder to parse the XMl.
+      val builderFactory = DocumentBuilderFactory.newInstance()
+
+      // document builder which parses the XMl.
+      val docBuilder = builderFactory.newDocumentBuilder()
+
+      docBuilder.setErrorHandler(SyntaxErrorHandler(repoPath))
+
+      // check if the repo has any syntactically incorrect XML.
+      val hasXmlSyntaxFailure = searchFiles.fold(false) { isFailing, file ->
+        val fileResult = xmlSyntaxCheck.checkIfFileHasBrokenXml(docBuilder, file)
+        isFailing || fileResult
+      }
+
+      if (hasXmlSyntaxFailure) {
+        throw Exception(ScriptResultConstants.XML_SYNTAX_CHECK_FAILED)
+      } else {
+        println(ScriptResultConstants.XML_SYNTAX_CHECK_PASSED)
+      }
+    }
+  }
+
+  /**
+   * Check if a file has a broken XML syntax.
+   *
+   * @param docBuilder the builder which will parse the XML file
+   * @param file the file to be checked for
+   * @return whether a file is syntactically correct or not
+   */
+  private fun checkIfFileHasBrokenXml(docBuilder: DocumentBuilder, file: File): Boolean {
+    try {
+      docBuilder.parse(file)
+    } catch (e: SAXParseException) {
```

Comment:
I wonder if maybe a cleaner way of handling this is to collect all the errors from the error handler, and if any are present then we determine it's a failure & log them. This keeps the logging & the logic in the same file since the error handler just becomes a cataloging utility, and it avoids using exceptions for control flow (which is considered a bad practice in Effective Java).

Reply:
This sounds neat to me. Here is what I am thinking, we will maintain a mutable list of errors in the `XmlSyntaxErrorHandler`, and whenever any exception is raised by the parser, we will simply add it to the list. After the parsing is done, we will just collect the list of all the errors received by the error handler and then log them and fail the script if it's the case. How does this sound?

Also, one thing that we have to take care of is that, for every `fatalError` in the Xml syntax, if we are throwing no exception in the error handler, then the default error handler throws a `SaxParseException`. So, inorder to make our script continue to work without being terminated in between, we have to wrap the parsing line of code in a try-catch block, just like this:
```kotlin
try {
 builder.parse('path_to_xml')
} catch (e: SaxParseException){}
```
Please let me know your thoughts on this overall approach.

Reply:
Both sound good to me, and I agree with the overall approach being proposed here.

Reply:
Implemented.